build.gradle

```diff
@@ -93,123 +91,126 @@ class PackageProperties {
+        jreProperties.findAll {String sys, JreProperties props -> !props.jre.exists()}
+                .collect {String sys, JreProperties props ->
+            Thread.start {
+                println("Start downloading OpenJRE for ${sys}...")
+                new URL(props.downloadUrl as String).withInputStream {
+                    props.jre.newOutputStream() << it
+                }
+                if (checkHash(props)) {
+                    println("OpenJRE for ${sys} is downloaded.")
+                } else {
+                    println("Hash mismatch for downloaded ${sys} JRE. File may have been modified.")
+                    System.exit(1)
                 }
             }
-        }.each {
-            thread -> thread.join()
-        }
+        }*.join()
     }
 }
 
-task unzipJre(description: 'Unzip JREs only when the extract directory is not existing', dependsOn:downloadJre) {
-    doFirst {
-        fileTree(jreDownloadDir).matching {
-            // Excluding the jre folder
-            exclude 'jre'
-        }.each {jreFile ->
-             if(jreFile.name.endsWith(ZIP)) {
-                 copy {
-                     from zipTree(jreFile)
-                     into "$jreExtractDir/${WINDOWS}"
-                 }
-             }
-             else {
-                 copy {
-                     from tarTree(resources.gzip(jreFile))
-                     into "$jreExtractDir/${jreFile.name.contains(MAC) ? MAC : LINUX}"
-                 }
-             }
-        }
+// Execute downloadJre only when any one of three JREs is missing
+downloadJre.onlyIf {
+    jreProperties.any {sys,props ->
+        !props.jre.exists()
     }
 }
 
-unzipJre.onlyIf {
-    !file(jreExtractDir).exists()
-}
+static boolean checkHash(JreProperties props) {
+    final hash = MessageDigest.getInstance('SHA-256').with {
+        props.jre.eachByte(1024) {buff,len ->
+            update(buff, 0, len)
+        }
+        digest().encodeHex() as String
+    }
 
-task createPackages(description: 'Create three packages for Linux, Windows and Mac', dependsOn:unzipJre) {
-    //This task also depends on other three tasks which are created by iterating osNames
+    hash == props.hash
 }
 
-packagePropertiesMap.each {String sys, PackageProperties props ->
-    props.with {
-        String scriptDir = "${projectDir}/launcher-scripts/${scriptFile}"
-        String jreExtractSubDir = "${jreExtractDir}/$sys"
-        Class taskCLass = Class.forName("org.gradle.api.tasks.bundling.${taskType}")
-
-        task "packageFor$sys"(type: taskCLass, description: "Create ${sys} package") {
-            mustRunAfter unzipJre
-            archiveBaseName = "admin-console-package-for-${sys}"
-            extension = packageFileExt
-            if(fileCompression) {
-                compression = Compression.GZIP
+task unzipJre(description: 'Unzip a JRE when the extract directory is missing', dependsOn:downloadJre) {
+    doFirst {
+
```

Comment:
I will remove this line.
Also I wonder if we still want to set when execute this task by 'onlyif'.